packages/react-async/src/helpers.tsx

```diff
@@ -0,0 +1,138 @@
+import React from "react"
+import propTypes from "./propTypes"
+
+import {
+  InitialChildren,
+  PendingChildren,
+  FulfilledChildren,
+  RejectedChildren,
+  SettledChildren,
+  AsyncState,
+  AbstractState,
+} from "./Async"
+
+/**
+ * Due to https://github.com/microsoft/web-build-tools/issues/1050, we need
+ * AbstractState imported in this file, even though it is only used implicitly.
+ * This _uses_ AbstractState so it is not accidentally removed by someone.
+ */
+declare type ImportWorkaround<T> = AbstractState<T>
+
+// these were exported as type, but never existed
+// export declare function IfLoading<T>(props: {
+//   children?: PendingChildren<T>
+//   initial?: boolean
+//   state: AsyncState<T>
+// }): React.ReactNode
+// export declare function IfResolved<T>(props: {
+//   children?: FulfilledChildren<T>
+//   persist?: boolean
+//   state: AsyncState<T>
+// }): React.ReactNode
+
+const nullify = <T extends {}>(children: T | undefined): T | null =>
+  children === undefined ? null : children
+
+const renderFn = (
+  children: React.ReactNode | ((...args: any[]) => React.ReactNode),
+  ...args: any[]
+) => nullify(typeof children === "function" ? (children as any)(...args) : children)
+
+/**
+ * Renders only when no promise has started or completed yet.
+ *
+ * @prop {Function|Node} children Function (passing state) or React node
+ * @prop {Object} state React Async state object
+ * @prop {boolean} persist Show until we have data, even while pending (loading) or when an error occurred
+ */
+export const IfInitial = <T extends {}>({
+  children,
+  persist,
+  state = {} as any,
+}: {
+  children?: InitialChildren<T>
+  persist?: boolean
+  state: AsyncState<T>
+}) => <>{state.isInitial || (persist && !state.data) ? renderFn(children, state) : null}</>
+
+/**
+ * Renders only while pending (promise is loading).
+ *
+ * @prop {Function|Node} children Function (passing state) or React node
+ * @prop {Object} state React Async state object
+ * @prop {boolean} initial Show only on initial load (data is undefined)
+ */
+export const IfPending = <T extends {}>({
+  children,
+  initial,
+  state = {} as any,
+}: {
+  children?: PendingChildren<T>
+  initial?: boolean
+  state: AsyncState<T>
+}) => <>{state.isPending && (!initial || !state.value) ? renderFn(children, state) : null}</>
```

Comment:
Yup, otherwise TS doesn't recognize the return value correctly (not every path returns a JSX.Element). And we can't type it as React.FC because it's generic.